#  **Price Analysis**

Get the data from the csv file

In [ ]:
#Let's import useful libraries

import pandas as pd
import numpy as np
import re

In [ ]:
df = pd.read_csv('preprocessed_data_craigslist.csv')
df.drop(columns = 'Unnamed: 0', axis = 1, inplace = True)
df.head()

price location                                                url  \
0  1000.0   auburn  https://auburn.craigslist.org/mcd/d/fayettevil...   
1     1.0   auburn  https://auburn.craigslist.org/for/d/defuniak-s...   
2  8795.0   auburn  https://auburn.craigslist.org/fod/d/fairburn-2...   
3    30.0   auburn  https://auburn.craigslist.org/bop/d/opelika-th...   
4   112.0   auburn  https://auburn.craigslist.org/sgd/d/full-garag...   

               date                                              title  \
0  2022-04-05 15:32  BAD CREDIT, NO CREDIT, OK! WE WORK WITH EVERYONE!   
1  2022-04-17 13:32                                              Atv’s   
2  2022-04-16 14:51  2017 Club Car Precedent 4 Seater Gas Alabama C...   
3  2022-04-16 09:07  Thule 961XT Speedway Bike Strap Rear Rack Carrier   
4  2022-04-15 12:52  Full Garage Gym-Squat Rack, Dumbbells- Financi...   

   numimage                                               text condition  \
0      10.0  WE SHIP NATIONWIDE, FINANCE NATIONWIDE! YOU SE...   unknown   
1      23.0  Used and New atv’s , go karts and dirt bikes f...   unknown   
2      20.0  Very Nice 2017 Club Car Precedent EFI Gas Alab...         4   
3       6.0  Thule 961XT Speedway - Bike Strap Rear Rack Ca...   unknown   
4       4.0  Full Garage Gym Setup- Squat Rack, Adjustable ...         4   

   makemanufacturer  modelnamenumber  ... electricassist framesize  \
0               0.0              0.0  ...            0.0   unknown   
1               0.0              0.0  ...            0.0   unknown   
2               0.0              0.0  ...            0.0   unknown   
3               1.0              1.0  ...            0.0   unknown   
4               1.0              1.0  ...            0.0   unknown   

  handlebartype  suspension wheelsize sizedimensions  serialnumber paintcolor  \
0       unknown         0.0   unknown              0             0          0   
1       unknown         0.0   unknown              0             0          0   
2       unknown         0.0   unknown              0             0          0   
3       unknown         0.0   unknown              0             0          0   
4       unknown         0.0   unknown              0             0          0   

   yearmanufactured  days  
0                 0    16  
1                 0     4  
2                 0     5  
3                 0     5  
4                 0     6  

[5 rows x 23 columns]

In [ ]:
df.price.describe()

count      3437.000000
mean       4513.562118
std       24026.705163
min           1.000000
25%          75.000000
50%         275.000000
75%        2295.000000
max      999992.000000
Name: price, dtype: float64

We observe that there is a huge standard deviation for the price distribution. The minimum of $1 corresponds to annouces where the owner had to make an input in the price category to publish the annouce but wants to discuss the price. We can imagine the same for the prices very high. We could get some information about the price in the description. That is what we will look at in the next section.

Let's filter the data to get the annouces with incoherent prices. We have taken arbitrary threshold to determine if an announce is incoherent or not. 

In [ ]:
df_filter = df[(df['price']>3000) | (df['price']<10)]
print(f'The number of incoherent announces regarding their price is: ',len(df_filter))
df_filter

The number of incoherent announces regarding their price is:  922


price location                                                url  \
1        1.0   auburn  https://auburn.craigslist.org/for/d/defuniak-s...   
2     8795.0   auburn  https://auburn.craigslist.org/fod/d/fairburn-2...   
9        1.0   auburn  https://bham.craigslist.org/bik/d/springville-...   
27    4800.0   auburn  https://atlanta.craigslist.org/sat/mcy/d/lagra...   
40       1.0   auburn  https://atlanta.craigslist.org/atl/bik/d/atlan...   
...      ...      ...                                                ...   
3355     1.0  seattle  https://seattle.craigslist.org/sno/bik/d/lynnw...   
3356  3750.0  seattle  https://seattle.craigslist.org/skc/mpo/d/black...   
3382     1.0  seattle  https://seattle.craigslist.org/see/wan/d/seatt...   
3418     5.0  seattle  https://seattle.craigslist.org/oly/tag/d/olymp...   
3423     1.0  seattle  https://seattle.craigslist.org/see/bik/d/seatt...   

                  date                                              title  \
1     2022-04-17 13:32                                              Atv’s   
2     2022-04-16 14:51  2017 Club Car Precedent 4 Seater Gas Alabama C...   
9     2022-04-18 09:36                    Mountain bikes / children bikes   
27    2022-04-15 19:58                           2009/2014 Kawasaki Bikes   
40    2022-04-05 08:38                            Multiple bikes for sale   
...                ...                                                ...   
3355  2022-04-18 12:15                          6 Ladies bikes-$50 to $75   
3356  2022-04-18 11:34                                     KTM Dirt Bikes   
3382  2022-04-15 10:52            CASH for wrecked or damaged sport bikes   
3418  2022-04-04 11:06           Motormax Super Bikes Kawasaki 1/24 scale   
3423  2022-03-31 11:30                              SCHWINN CLASSIC BIKES   

      numimage                                               text condition  \
1         23.0  Used and New atv’s , go karts and dirt bikes f...   unknown   
2         20.0  Very Nice 2017 Club Car Precedent EFI Gas Alab...         4   
9         18.0  I have a Vertical PK7 21 speed mountain bike w...   unknown   
27        12.0  Blue-2009. Asking 3800.00 Damage on light. No ...   unknown   
40         3.0  multiple bikes for sale -  cannondale R600 (60...   unknown   
...        ...                                                ...       ...   
3355       0.0  Roadmaster Mountain Sport, 18 speed, 24" tires...         3   
3356       9.0  2017 KTM 85 SX less than 5 hrs  $5000  2012 KT...   unknown   
3382       0.0  I am in the market to purchase wrecked or dama...   unknown   
3418       0.0  Motormax Super Bikes Kawasaki motorcycle   I c...         4   
3423       0.0  SCHWINN CLASSIC BIKES, SELLING AS A PAIR. ONLY...   unknown   

      makemanufacturer  modelnamenumber  ... electricassist framesize  \
1                  0.0              0.0  ...            0.0   unknown   
2                  0.0              0.0  ...            0.0   unknown   
9                  0.0              0.0  ...            0.0         3   
27                 0.0              0.0  ...            0.0   unknown   
40                 0.0              0.0  ...            0.0         3   
...                ...              ...  ...            ...       ...   
3355               1.0              0.0  ...            0.0   unknown   
3356               1.0              1.0  ...            0.0   unknown   
3382               0.0              0.0  ...            0.0   unknown   
3418               0.0              0.0  ...            0.0   unknown   
3423               0.0              0.0  ...            0.0         1   

     handlebartype  suspension wheelsize sizedimensions  serialnumber  \
1          unknown         0.0   unknown              0             0   
2          unknown         0.0   unknown              0             0   
9          unknown         0.0   unknown              0             0   
27         unknown         0.0   unknown           

In [ ]:
print(f'The price for this announce on Craigslist was: $', df_filter['price'][3423] )
print(f'Here is the description associated with this annouce: ',df_filter['text'][3423])

The price for this announce on Craigslist was: $ 1.0
Here is the description associated with this annouce:  SCHWINN CLASSIC BIKES, SELLING AS A PAIR. ONLY $100 FOR BOTH!


Let's use a regular expression pattern to extract the price from the annouce description. 

In [ ]:
from statistics import mean

index_list = df_filter.index
index_list_final = list()
new_price_list = list()

for i in df.index:
  if i in df_filter.index:
    new_price_int = list()
    try:
      new_price = re.findall('\$(\x20?\d+(?:[,]\d{0,3})?)', df_filter['text'][i])
      for j in new_price:
        new_price_int.append(int(j.replace(',','')))
      if new_price_int == []:
        new_price = float('Nan')
        new_price_list.append(new_price)
        index_list_final.append(i)
      else:
        new_price_list.append(int(sum(new_price_int)/len(new_price_int)))
        index_list_final.append(i)
    except:
      #If there is no match, we replace the price value by NaN to eliminate them easily from our database
      new_price_list.append(float('Nan'))
      index_list_final.append(i)
  else:
    new_price_list.append(df['price'][i])
    index_list_final.append(i)

df_new_price = pd.DataFrame({'price' : new_price_list, 'Index':index_list_final})
df_new_price.set_index('Index', inplace = True)
df_new_price

price
Index        
0      1000.0
1         NaN
2      2995.0
3        30.0
4       112.0
...       ...
3432     75.0
3433    100.0
3434    110.0
3435     50.0
3436    375.0

[3437 rows x 1 columns]

In [ ]:
#We update our dataframe with the new prices

df['price'] = df_new_price['price']
df.head()

price location                                                url  \
0  1000.0   auburn  https://auburn.craigslist.org/mcd/d/fayettevil...   
1     NaN   auburn  https://auburn.craigslist.org/for/d/defuniak-s...   
2  2995.0   auburn  https://auburn.craigslist.org/fod/d/fairburn-2...   
3    30.0   auburn  https://auburn.craigslist.org/bop/d/opelika-th...   
4   112.0   auburn  https://auburn.craigslist.org/sgd/d/full-garag...   

               date                                              title  \
0  2022-04-05 15:32  BAD CREDIT, NO CREDIT, OK! WE WORK WITH EVERYONE!   
1  2022-04-17 13:32                                              Atv’s   
2  2022-04-16 14:51  2017 Club Car Precedent 4 Seater Gas Alabama C...   
3  2022-04-16 09:07  Thule 961XT Speedway Bike Strap Rear Rack Carrier   
4  2022-04-15 12:52  Full Garage Gym-Squat Rack, Dumbbells- Financi...   

   numimage                                               text condition  \
0      10.0  WE SHIP NATIONWIDE, FINANCE NATIONWIDE! YOU SE...   unknown   
1      23.0  Used and New atv’s , go karts and dirt bikes f...   unknown   
2      20.0  Very Nice 2017 Club Car Precedent EFI Gas Alab...         4   
3       6.0  Thule 961XT Speedway - Bike Strap Rear Rack Ca...   unknown   
4       4.0  Full Garage Gym Setup- Squat Rack, Adjustable ...         4   

   makemanufacturer  modelnamenumber  ... electricassist framesize  \
0               0.0              0.0  ...            0.0   unknown   
1               0.0              0.0  ...            0.0   unknown   
2               0.0              0.0  ...            0.0   unknown   
3               1.0              1.0  ...            0.0   unknown   
4               1.0              1.0  ...            0.0   unknown   

  handlebartype  suspension wheelsize sizedimensions  serialnumber paintcolor  \
0       unknown         0.0   unknown              0             0          0   
1       unknown         0.0   unknown              0             0          0   
2       unknown         0.0   unknown              0             0          0   
3       unknown         0.0   unknown              0             0          0   
4       unknown         0.0   unknown              0             0          0   

   yearmanufactured  days  
0                 0    16  
1                 0     4  
2                 0     5  
3                 0     5  
4                 0     6  

[5 rows x 23 columns]

In [ ]:
df.price.describe()

count      2944.000000
mean       1632.134511
std        9448.964930
min           0.000000
25%          80.000000
50%         225.000000
75%        1000.000000
max      317422.000000
Name: price, dtype: float64

We observe that the mean is now 1632 against 4513 before the cleaning of our data. 
This is much more coherent with the price of bikes.

In [ ]:
#Let's check how many useful data we still have in our dataframe

df.price.count() 

2944